### ISSUE TITLE
Updating existing extensions

### ISSUE THREAD

**Issue by gintuoklis** [HodorNV/ALOps#73]
hey Guys, 
I am using ALOpsAppPublish in release pipeline to update TEST environment with latest extensions. Release works fine for the first. Second time updating the same extension it gives an error regarding different version:

`Cannot install the extension <ExtensionName> by <Publisher> 1.0.202004.32805 because the tenant default already uses a different version of it.`

If i check test environment i see two records in Extension Management:

![image](https://user-images.githubusercontent.com/16342128/73213049-ca459100-414f-11ea-894d-d8d50b6afee1.png)

Am i doing something wrong? Shouldn't update existing extensin with a new version?

This is the yaml for publish task":

```
steps:
- task: Hodor.hodor-alops.alopspublishtask.ALOpsAppPublish@1
  displayName: 'ALOps App Publish'
  inputs:
    nav_serverinstance: NAV
    batch_publish_folder: '$(System.ArtifactsDirectory)'
```

Best Regards, 
Gintautas

**Comment by waldo1001** [comment]
Well, that's normal behavior.  If you install the same app a second time when it's already there, it refuses that for that reason.  You can only upgrade to a new version.  It doesn't make sens to publish the same app twice, does it ;-).

You need to make sure you release from a new build from a build pipeline that manages the version of the app - if so, every build has a new artifact with a new (higher) version.  

Example:
```yaml
- task: ALOpsAppCompiler@1
  displayName: 'ALOps Compile Extension: App'
  inputs:
    usedocker: true
    nav_app_version: '1.0.[yyyyWW].*'
    targetproject: 'app/app.json' 
    failed_on_warnings: true  
```

How does your build look like?  Can you share the yaml?

**Comment by gintuoklis** [comment]
Well, i am not publishing the same app. Each time build finishes it increments app version. 

My build pipeline pipeline looks like this:

```
steps:
- checkout: self
  clean: true 
  
- task: Docker@2
  displayName: Login to ACR
  inputs:
   command: login
   containerRegistry: DevopsRegistryServiceConnection

- task: ALOpsDockerStart@1
  inputs:
    docker_image: 'mydevops.azurecr.io/imv:latest'
    docker_pull: true
    enable_symbol_loading: true
    accept_image_outdated: true 

- task: ALOpsDockerWait@1
  inputs:
    search_string: 'Ready for connections!'

- task: ALOpsAppCompiler@1
  displayName: 'Compile Extension'
  inputs:
    usedocker: true
    nav_app_version: '1.0.[yyyyWW].*'
    failed_on_warnings: true    

- task: ALOpsAppPublish@1
  displayName: 'Publish AL-App'
  inputs:
    usedocker: true
    nav_artifact_app_filter: '*.app'   
    skip_verification: true 

- task: ALOpsAppTest@1
  displayName: 'Run TestSuite'
  inputs:
    usedocker: true
    testfilter: '50000..99999'
    import_testtoolkit: false
    show_available_tests: true

  continueOnError: true

- task: PublishTestResults@2
  displayName: 'Publish Test Results **/TestResults.xml'
  inputs:
    testResultsFormat: XUnit
    testResultsFiles: '**/TestResults.xml'
    testRunTitle: 'BC Test Results: $(Build.BuildId)'

- task: ALOpsDockerRemove@1
  displayName: 'Remove Docker Container'
  enabled: true
  condition: always()

```
This produce artifacts that i use later in release pipeline. Release pipeline has only one task to publishextension on local server

```
steps:
- task: Hodor.hodor-alops.alopspublishtask.ALOpsAppPublish@1
  displayName: 'ALOps App Publish'
  inputs:
    nav_serverinstance: NAV
    batch_publish_folder: '$(System.ArtifactsDirectory)'


```

**Comment by gintuoklis** [comment]
If i have only one extension per release it seems to work fine. But once, I try to publish multiple extensions it starts failing. Let's say i have 10 extensions linked to my release pipeline. I updated just one of them. All other remain with the same version. Nevertheless, release will try to publish them anyhow. The ones with the same appversion, just generate warnings and succeeds. Extension which changed app version, which produce the error i talked in the first post. 

Perhpa that helps for the investigation? Perhaps it's just an issue with a new parameter _bathc_publish_folder_?

**Comment by AdminHodor** [comment]
Hello @gintuoklis ,

The batch publish does indeed not include the same array of checks as for a single publish. 
We'll look into this and get this on the roadmap.

Kind regards

**Comment by gintuoklis** [comment]
Hey @AdminHodor, 
What's the purpose of _batch_publish_folder_  then? How do you use it in practise?

**Comment by AdminHodor** [comment]
Dear @gintuoklis ,

Could you please validate again? We changed the behavior of batch_publish so it works the same as the normal publish. Also depencies are detected and installed in proper order.

Kind regards,

**Comment by gintuoklis** [comment]
Sorry, cannot get it working at all now.

My task look likes this: 
```
- task: ALOpsAppPublish@1
  displayName: 'Publish AL-App'
  inputs:
    usedocker: false
    nav_artifact_app_filter: '*.app'   
    skip_verification: true
    nav_serverinstance: ${{ parameters.instanceName }}
    batch_publish_folder: $(Pipeline.Workspace)\extensions\
```

It gives me this error:
`##[error]The path cannot be processed because it resolved to more than one file; only one file at a time can be processed.
`
extension folder looks like this:
![image](https://user-images.githubusercontent.com/16342128/78510050-c5760c80-7792-11ea-8756-1e5d95a90a38.png)

I have a feeling that it cannot resolve the path at all. 
I even tried to move the contents to C:\temp\extensions\, and hardcode this path into the parameters. But the result was the same - the same error message.

**Comment by AdminHodor** [comment]
Dear @gintuoklis ,

>>Apparently, this parameter is not working at all.
Doubtfully, we have (and seen) this running in multiple pipelines.

Could you please provide the output of the step? 

Kind regards,

Bart Bourgeois

**Comment by gintuoklis** [comment]
Here's the full output. 
There's something ill with this step. It complains about resolving multiple extensions files, while the whole point of batch publishing is to publish multiple extensions. 

```
Starting: Publish AL-App
==============================================================================
Task         : ALOps App Publish
Description  : Publish NAV Apps
Version      : 1.428.1179
Author       : Hodor
Help         : Publish Business Central extension to service tier.
==============================================================================
*** Validate configuration
*** Importing required PS-Functions
 - Use Docker         = False
 - NAV ServerInstance = buildTest
 - Artifact Path      = 
 - App Filter         = *.app
 - Process Type       = Build
 - Skip Verification  = True

*** Importing required PS-Functions
*** Start Batch App-Publish.
##[error]The path cannot be processed because it resolved to more than one file; only one file at a time can be processed.
Finishing: Publish AL-App
```

**Comment by AdminHodor** [comment]
Dear @gintuoklis ,

This seems exactly the same issue as #104 , cfr having multiple versions of NAV/BC installed.
From the log I can see it fails even before the publish even starts.

Kind regards,

**Comment by AdminHodor** [comment]
Dear @gintuoklis ,

We made a hotfix for the Multiple ServiceTier installations, which should address this issue.
Hotfix is included in release v1.428.1187.

Kind regards,
